sklearn/impute.py

```diff
@@ -362,6 +379,26 @@ def _dense_fit(self, X, strategy, missing_values, fill_value):
         # Constant
         elif strategy == "constant":
             return np.full(X.shape[1], fill_value, dtype=X.dtype)
+    
+        # Sample
+        elif strategy == "sample":
+            X = X.transpose()
+            mask = mask.transpose()
+
+            generators = np.empty(X.shape[0], dtype=object)
+
+            for i, (row, row_mask) in enumerate(zip(X[:], mask[:])):
+                row_mask = np.logical_not(row_mask).astype(np.bool)
+                row = row[row_mask]
+                if row.size > 0:
+                    uniques, counts = np.unique(row, return_counts=True)
+                    probas = counts / counts.sum()
+                    g = lambda k: random_state.choice(uniques, k, p=probas)
+                    generators[i] = g
+                else:
+                    generators[i] = np.nan
+
+            return generators
```

Comment:
I don't like this overloading of `statistics_` with something of a very different meaning.

Reply:
I guess since we can't store a closure as an attribute, we have to directly store the non-missing values array and their probas. Then store it in ``statistics_`` would make sens since we would store the distribution itself.